modules/objdetect/src/qrcode.cpp

```diff
@@ -3816,33 +3835,469 @@ bool QRDetectMulti::computeTransformationPoints(const size_t cur_ind)
+            int id = (innerCornerId + 3) % 4;
+            return std::make_pair(id, points[id]);
+        }
+        else if (typePattern == TypePattern::BOTTOM) {
+            int id = (innerCornerId + 1) % 4;
+            return std::make_pair(id, points[id]);
+        }
+        return std::make_pair(-1, Point2f());
+    }
+
+    Point2f getTimingStart(TypePattern direction) const {
+        Point2f start = points[innerCornerId]*(6.5f/7.f);
+        if (typePattern == TypePattern::CENTER && direction == TypePattern::RIGHT) {
+            start += points[(innerCornerId + 3) % 4]*(0.5f/7.f);
+        }
+        else if (typePattern == TypePattern::CENTER && direction == TypePattern::BOTTOM) {
+            start += points[(innerCornerId + 1) % 4]*(0.5f/7.f);
+        }
+        else if (typePattern == TypePattern::RIGHT && direction == TypePattern::CENTER) {
+            start += points[(innerCornerId + 1) % 4]*(0.5f/7.f);
+        }
+        else if (typePattern == TypePattern::BOTTOM && direction == TypePattern::CENTER) {
+            start += points[(innerCornerId + 3) % 4]*(0.5f/7.f);
+        }
+        return start + getDirectionTo(direction)/(7.f*2.f);
+    }
+
+    // return (total white+black pins, total first color, penaltyPoints)
+    Point3i calculateTimingPattern(const Point2f& start, const Point2f& end, Mat &img, const float maxTimingPatternDiff) const {
+        Rect imageRect(Point(), img.size());
+        int penaltyPoints = 0;
+        int colorCounters[2] = {0, 0};
+        if (imageRect.contains(Point(cvRound(end.x), cvRound(end.y)))) {
+            LineIterator lineIterator(start, end);
+            uint8_t prevValue = img.at<uint8_t>(lineIterator.pos());
+
+            vector<Point> vec = {lineIterator.pos()};
+
+            bool firstColor = true;
+            lineIterator++;
+            colorCounters[firstColor]++;
+
+            for(int i = 1; i < lineIterator.count; i++, ++lineIterator) {
+                const uint8_t value = img.at<uint8_t>(lineIterator.pos());
+                if (prevValue != value) {
+                    const float dist = sqrt(normL2Sqr<float>((Point2f)(vec.back()-lineIterator.pos())));
+                    // check long and short lines in timing pattern
+                    const float relativeDif = max(moduleSize, dist)/min(moduleSize, dist);
+                    if (relativeDif > maxTimingPatternDiff) {
+                        if (dist < moduleSize || relativeDif < maxTimingPatternDiff*8.f)
+                            penaltyPoints++;
+                        else
+                            penaltyPoints += cvRound(relativeDif);
+                    }
+                    vec.push_back(lineIterator.pos());
+                    prevValue = value;
+                    firstColor ^= true;
+                    colorCounters[firstColor]++;
+                }
+            }
+        }
+        // penaltyPoints could be is 2, because start and end in in the center of the module
```

Comment:
Not clear comment.

Reply:
removed comment, the comment has lost its relevance.